### PR TITLE
✨ Bump CAPI to v1.3.0

### LINKED ISSUE #1315: Migrate to Ginkgo v2 during cluster-api v1.3 bump
/kind feature

**Describe the solution you'd like**

Starting from CAPI v1.3, Ginkgo v2 will be used (https://github.com/kubernetes-sigs/cluster-api/pull/6906).
For bumping to CAPI 1.3 version, we should also migrate e2e tests to v2.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

### PR DESCRIPTION
**Pull request by tobiasgiese**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1315

This PR bumps the CAPI version to v1.3.0 and respects the [version migration guide](https://cluster-api.sigs.k8s.io/developer/providers/v1.2-to-v1.3.html).

Non-obvious ginkgo v2 changes:

- updated `setup-envtest` (i.e., controller-runtime) in `Makefile` to the most recent `master`.
  Otherwise, the `etcd` and `kube-apiserver` binaries are not available in our unit tests with ginkgo v2
- changed deprecated ginkgo parameter `-progress` to `-show-node-events`
- [move `By()` inside `It()`](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/1406/files#diff-c611c23bcd3d689e7b0d56720f5b099a9f7bfa59d06a91cfc48a14cfeb042c71) to fix ginkgo errors (see [Documenting Complex Specs: By](https://onsi.github.io/ginkgo/#documenting-complex-specs-by))
- Added `-timeout=3h` to the e2e tests because [the default timeout has been reduced from 24h down to 1h](https://onsi.github.io/ginkgo/MIGRATING_TO_V2#timeout-behavior)

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**TODOs**:

- [x] squashed commits
- if necessary:
  - [ ] includes documentation
  - [ ] adds unit tests

/hold

### CODE REVIEW

**Comment by tobiasgiese** [comment]
> Waiting for #1408 to be merged

e2e tests are green

**Comment by tobiasgiese** [comment]
Do we want to add CAPI v1.3.0 to our v0.7 release? WDYT @mdbooth?

**Comment by tobiasgiese** [comment]
Feel free to remove the hold @mdbooth

File: test/e2e/data/kubetest/conformance-fast.yaml [hunk 1]
```diff
@@ -1,8 +1,9 @@
 ginkgo.focus: \[Conformance\]
 ginkgo.skip: .*\[Serial\].*
 disable-log-dump: true
-ginkgo.progress: true
+ginkgo.show-node-evens: true
```
Comment (furkatgofurov7) [inline test/e2e/data/kubetest/conformance-fast.yaml:4]: nit: bumped into this before and here accidentally, spelling is incorrect for the new parameter, should be `--show-node-events`, see the deprecation warning below for ref: 
```
 Test Suite Failed
You're using deprecated Ginkgo functionality:
=============================================
  --progress is deprecated .  The functionality provided by --progress was confusing and is no longer needed.  Use --show-node-events instead to see node entry and exit events included in the timeline of failed and verbose specs.  Or you can run with -vv to always see all node events.  Lastly, --poll-progress-after and the PollProgressAfter decorator now provide a better mechanism for debugging specs that tend to get stuck. 
```

Comment (tobiasgiese) [reply test/e2e/data/kubetest/conformance-fast.yaml]: Interesting copy&paste fail. Even in the PR description I've added the wrong parameter 😅 🤷🏻‍♂️

Comment (tobiasgiese) [reply test/e2e/data/kubetest/conformance-fast.yaml]: Also forgot to change `-progress` inside our `Makefile` -- initially searched for `--progress` 🤦🏻‍♂️

File: test/e2e/suites/conformance/conformance_test.go [hunk 2]
```diff
@@ -66,15 +64,6 @@ var _ = Describe("conformance tests", func() {
...
 
-		// Starting with Kubernetes v1.25, the kubetest config file needs to be compatible with Ginkgo V2.
-		v125 := semver.MustParse("1.25.0-alpha.0.0")
-		v, err := semver.ParseTolerant(kubernetesVersion)
-		Expect(err).NotTo(HaveOccurred())
-		if v.GTE(v125) {
-			// Use the Ginkgo V2 config file.
-			e2eCtx.Settings.KubetestConfigFilePath = getGinkgoV2ConfigFilePath(e2eCtx.Settings.KubetestConfigFilePath)
-		}
-
```
Comment (mdbooth) [inline test/e2e/suites/conformance/conformance_test.go:77]: Not worth re-doing now, but when we're reverting a commit can we do it in a separate commit? i.e. The outcome of `git revert`? It just makes it obvious exactly why a change was made both during review and when spelunking in git in the future without having to refer to the PR.

**Review by mdbooth — COMMENTED** [review]
Looks good.

I'm confused about the inclusion of E2E_IMAGE_URL, though. Was this broken before? If not, what is included in this PR which requires it?

File: test/e2e/data/e2e_conf.yaml [hunk 3]
```diff
@@ -137,6 +137,7 @@ variables:
   CONFORMANCE_WORKER_MACHINE_COUNT: "5"
   CONFORMANCE_CONTROL_PLANE_MACHINE_COUNT: "1"
   INIT_WITH_KUBERNETES_VERSION: "v1.25.0"
+  E2E_IMAGE_URL: "http://10.0.3.15/capo-e2e-image.tar"
```
[the remark this answers is not in the document]
Comment (tobiasgiese) [reply test/e2e/data/e2e_conf.yaml:140]: This was moved from https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/31f7c9c9ab6191e4700ff41f1b8e53af29a2dd5f/test/e2e/suites/e2e/clusterctl_upgrade_test.go#L40
If we don't specify it in the `e2e_conf.yaml` the following error occurs

```
  [FAILED] Failed to run clusterctl config cluster
  Unexpected error:
      <*yamlprocessor.errMissingVariables | 0xc00121d740>: {
          Missing: ["E2E_IMAGE_URL"],
      }
      value for variables [E2E_IMAGE_URL] is not set. Please set the value using os environment variables or the clusterctl config file
  occurred
  In [It] at: /root/go/pkg/mod/sigs.k8s.io/cluster-api/test@v1.3.0/framework/clusterctl/client.go:302 @ 12/04/22 11:36:50.414
```

**Comment by tobiasgiese** [comment]
> I'm confused about the inclusion of E2E_IMAGE_URL, though. Was this broken before? If not, what is included in this PR which requires it?

This was just moved from the `clusterctl_upgrade_test.go` -- see https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/1406#discussion_r1042500735

File: Containerfile [hunk 4]
```diff
@@ -24,7 +24,7 @@
 #         media type: "application/vnd.oci.image.layer.v1.tar+gzip"
 
 # Build the manager binary
-FROM golang:1.18.3 as builder
```
[the remark this answers is not in the document]
Comment (tobiasgiese) [reply Containerfile:27]: Tbh, I totally forgot to bump the go version there as well 👍🏻 thanks

File: test/e2e/data/e2e_conf.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (tobiasgiese) [reply test/e2e/data/e2e_conf.yaml:140]: Prow job that failed with this error: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-openstack/1406/pull-cluster-api-provider-openstack-e2e-test/1599355701279330304

[the remark this answers is not in the document]
Comment (tobiasgiese) [reply test/e2e/data/e2e_conf.yaml:140]: Maybe @lentzi90 can say something about this?

[the remark this answers is not in the document]
Comment (mdbooth) [reply test/e2e/data/e2e_conf.yaml:140]: It looks like it was a bug in the previous PR: we're including `ci-hack-hc(p|t).yaml` unconditionally in all the E2E tests, and they reference `E2E_IMAGE_URL`. However, we're only setting the environment variable in the cluster upgrade test. Presumably the undefined variable error is a change of behaviour in CAPI 1.3, or we'd have noticed this before.

@lentzi90 I wonder if we want to be consistent in the use of `e2eCtx` here:
https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/31f7c9c9ab6191e4700ff41f1b8e53af29a2dd5f/test/e2e/suites/e2e/clusterctl_upgrade_test.go#L40-L41

Right now we're setting the same value in 2 places. It doesn't look like the clusterctl upgrade tests reads e2e_conf.yaml, so we likely can't remove the other one. I'd prefer to do this the same way in both places.

I don't think we need to hold up this PR, though. I'll open an issue to track it a tidy up.

[the remark this answers is not in the document]
Comment (mdbooth) [reply test/e2e/data/e2e_conf.yaml:140]: Ah, nm. It *is* initialised and you already fixed it 🤦

**Pull request merged**